FileNotFoundException causes app crash under unknown circumstances

Similar to issue #17, but occurs using version 4.0.0 and play store crash details do not show actual URI that caused FileNotFoundException.

I would recommend that a broad try/catch be inserted as far up the call stack as possible and a error Result be returned if any Exception is propagated up to that point. Otherwise an unknown Exception occurring in the plugin code can result in an app crash that won't be caught by the plugin caller from the Dart side of the channel.

Below is the full stacktrace provided by the play store crash reports for this issue.
```
java.lang.RuntimeException: 
  at android.os.AsyncTask$3.done (AsyncTask.java:300)
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:355)
  at java.util.concurrent.FutureTask.setException (FutureTask.java:222)
  at java.util.concurrent.FutureTask.run (FutureTask.java:242)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1112)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:587)
  at java.lang.Thread.run (Thread.java:841)
Caused by: java.io.FileNotFoundException: 
  at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel (DatabaseUtils.java:146)
  at android.content.ContentProviderProxy.openTypedAssetFile (ContentProviderProxy.java:682)
  at android.content.ContentResolver.openTypedAssetFileDescriptor (ContentResolver.java:1067)
  at android.content.ContentResolver.openAssetFileDescriptor (ContentResolver.java:907)
  at android.content.ContentResolver.openInputStream (ContentResolver.java:632)
  at com.sidlatau.flutterdocumentpicker.FileCopyTaskLoader.copyToTemp (FileCopyTaskLoader.java:184)
  at com.sidlatau.flutterdocumentpicker.FileCopyTaskLoader.loadInBackground (FileCopyTaskLoader.java:167)
  at com.sidlatau.flutterdocumentpicker.FileCopyTaskLoader.loadInBackground (FileCopyTaskLoader.java:165)
  at android.content.AsyncTaskLoader.onLoadInBackground (AsyncTaskLoader.java:312)
  at android.content.AsyncTaskLoader$LoadTask.doInBackground (AsyncTaskLoader.java:69)
  at android.content.AsyncTaskLoader$LoadTask.doInBackground (AsyncTaskLoader.java:57)
  at android.os.AsyncTask$2.call (AsyncTask.java:288)
  at java.util.concurrent.FutureTask.run (FutureTask.java:237)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1112)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:587)
  at java.lang.Thread.run (Thread.java:841)
```

--- a/android/src/main/kotlin/com/sidlatau/flutterdocumentpicker/FlutterDocumentPickerDelegate.kt
+++ b/android/src/main/kotlin/com/sidlatau/flutterdocumentpicker/FlutterDocumentPickerDelegate.kt
@@ -24,7 +24,7 @@ private const val LOADER_FILE_COPY = 603
 
 class FlutterDocumentPickerDelegate(
     private val activity: Activity
-) : PluginRegistry.ActivityResultListener, LoaderManager.LoaderCallbacks<String> {
+) : PluginRegistry.ActivityResultListener, LoaderManager.LoaderCallbacks<FileCopyTaskLoaderResult> {
     private var channelResult: MethodChannel.Result? = null
     private var allowedFileExtensions: Array<String>? = null
     private var invalidFileNameSymbols: Array<String>? = null
@@ -91,18 +91,22 @@ class FlutterDocumentPickerDelegate(
         }
     }
 
-    override fun onCreateLoader(id: Int, args: Bundle): Loader<String> {
+    override fun onCreateLoader(id: Int, args: Bundle): Loader<FileCopyTaskLoaderResult> {
         val uri = args.getParcelable<Uri>(EXTRA_URI)!!
         val fileName = args.getString(EXTRA_FILENAME)!!
         return FileCopyTaskLoader(activity, uri, fileName)
     }
 
-    override fun onLoadFinished(loader: Loader<String>?, data: String?) {
-        channelResult?.success(data)
+    override fun onLoadFinished(loader: Loader<FileCopyTaskLoaderResult>?, data: FileCopyTaskLoaderResult) {
+        if(data.isSuccess()) {
+            channelResult?.success(data.result)
+        } else {
+            channelResult?.error("LOAD_FAILED", data.error.toString(), null)
+        }
         activity.loaderManager.destroyLoader(LOADER_FILE_COPY)
     }
 
-    override fun onLoaderReset(loader: Loader<String>?) {
+    override fun onLoaderReset(loader: Loader<FileCopyTaskLoaderResult>?) {
     }
 
     private fun getFileCopyParams(resultCode: Int, data: Intent?): FileCopyParams? {
@@ -162,9 +166,14 @@ class FlutterDocumentPickerDelegate(
 
 data class FileCopyParams(val uri: Uri, val fileName: String, val extension: String?)
 
-class FileCopyTaskLoader(context: Context, private val uri: Uri, private val fileName: String) : AsyncTaskLoader<String>(context) {
-    override fun loadInBackground(): String {
-        return copyToTemp(uri = uri, fileName = fileName)
+class FileCopyTaskLoader(context: Context, private val uri: Uri, private val fileName: String) : AsyncTaskLoader<FileCopyTaskLoaderResult>(context) {
+    override fun loadInBackground(): FileCopyTaskLoaderResult {
+        try {
+            return FileCopyTaskLoaderResult(copyToTemp(uri = uri, fileName = fileName))
+        } catch (e: Exception) {
+            Log.e(FlutterDocumentPickerPlugin.TAG, "handlePickFileResult", e)
+            return FileCopyTaskLoaderResult(e)
+        }
     }
 
     override fun onStartLoading() {
@@ -193,5 +202,24 @@ class FileCopyTaskLoader(context: Context, private val uri: Uri, private val fil
         }
 
         return file.absolutePath
+    }
+}
+
+class FileCopyTaskLoaderResult {
+    val result: String?
+    val error: Exception?
+
+    constructor(result: String) {
+        this.result = result
+        this.error = null;
+    }
+
+    constructor(error: Exception) {
+        this.result = null;
+        this.error = error;
+    }
+
+    fun isSuccess(): Boolean {
+        return error == null
     }
 }

